File: src/Alimer/Numerics/BoundingBox.cs
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

using System.Diagnostics.CodeAnalysis;
using System.Numerics;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Runtime.Serialization;

namespace Alimer.Numerics;

/// <summary>
/// Defines an axis-aligned box-shaped 3D volume.
/// </summary>
[DataContract]
[StructLayout(LayoutKind.Sequential, Pack = 4)]
public partial struct BoundingBox : IEquatable<BoundingBox>, IFormattable
{
    /// <summary>
    /// Specifies the total number of corners (8) in the BoundingBox.
    /// </summary>
    public const int CornerCount = 8;

    /// <summary>
    /// A <see cref="BoundingBox"/> which represents an empty space.
    /// </summary>
    public static BoundingBox Zero => new(Vector3.Zero, Vector3.Zero);

    /// <summary>
    /// A <see cref="BoundingBox"/> which represents an infinite space.
    /// </summary>
    public static BoundingBox Infinite => new(float.MaxValue, float.MaxValue, float.MaxValue, float.MinValue, float.MinValue, float.MinValue);

    /// <summary>
    /// The minimum point of the box.
    /// </summary>
    public Vector3 Min;

    /// <summary>
    /// The maximum point of the box.
    /// </summary>
    public Vector3 Max;

    /// <summary>
    /// Initializes a new instance of the <see cref="BoundingBox"/> struct.
    /// </summary>
    /// <param name="min">The minimum vertex of the bounding box.</param>
    /// <param name="max">The maximum vertex of the bounding box.</param>
    public BoundingBox(Vector3 min, Vector3 max)
    {
        Min = min;
        Max = max;
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="BoundingBox"/> struct.
    /// </summary>
    /// <param name="minX">The x coordinate of the minimum point of the bounding box.</param>
    /// <param name="minY">The y coordinate of the minimum point of the bounding box.</param>
    /// <param name="minZ">The z coordinate of the minimum point of the bounding box.</param>
    /// <param name="maxX">The x coordinate of the maximum point of the bounding box.</param>
    /// <param name="maxY">The y coordinate of the maximum point of the bounding box.</param>
    /// <param name="maxZ">The z coordinate of the maximum point of the bounding box.</param>
    public BoundingBox(float minX, float minY, float minZ, float maxX, float maxY, float maxZ)
    {
        Min = new Vector3(minX, minY, minZ);
        Max = new Vector3(maxX, maxY, maxZ);
    }

    /// <summary>
    /// Gets the center of this bouding box.
    /// </summary>
    public readonly Vector3 Center
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => (Min + Max) / 2;
    }

    /// <summary>
    /// Gets the extent of this bouding box.
    /// </summary>
    public readonly Vector3 Extent
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => (Max - Min) / 2;
    }

    /// <summary>
    /// Gets size  of this bouding box.
    /// </summary>
    public readonly Vector3 Size => Max - Min;

    /// <summary>
    /// Gets the width of the bounding box.
    /// </summary>
    public readonly float Width
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => Extent.X * 2.0f;
    }

    /// <summary>
    /// Gets the height of the bounding box.
    /// </summary>
    public readonly float Height
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => Extent.Y * 2.0f;
    }

    /// <summary>
    /// Gets the depth of the bounding box.
    /// </summary>
    public readonly float Depth
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => Extent.Z * 2.0f;
    }

    /// <summary>
    /// Gets the volume of the bounding box.
    /// </summary>
    public readonly float Volume
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get
        {
            Vector3 sides = Max - Min;
            return sides.X * sides.Y * sides.Z;
        }
    }

    /// <summary>
    /// Get the perimeter length.
    /// </summary>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public readonly float GetPerimeter()
    {
        Vector3 sides = Max - Min;
        return 4 * (sides.X + sides.Y + sides.Z);
    }

    /// <summary>
    /// Get the surface area.
    /// </summary>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public readonly float GetSurfaceArea()
    {
        Vector3 sides = Max - Min;
        return 2 * (sides.X * sides.Y + sides.X * sides.Z + sides.Y * sides.Z);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public readonly void GetCorners(Span<Vector3> corners)
    {
        if (corners.Length < CornerCount)
        {
            throw new ArgumentOutOfRangeException(nameof(corners), $"GetCorners need at least {CornerCount} elements to copy corners.");
        }

        corners[0] = new Vector3(Min.X, Max.Y, Max.Z);
        corners[1] = new Vector3(Max.X, Max.Y, Max.Z);
        corners[2] = new Vector3(Max.X, Min.Y, Max.Z);
        corners[3] = new Vector3(Min.X, Min.Y, Max.Z);
        corners[4] = new Vector3(Min.X, Max.Y, Min.Z);
        corners[5] = new Vector3(Max.X, Max.Y, Min.Z);
        corners[6] = new Vector3(Max.X, Min.Y, Min.Z);
        corners[7] = new Vector3(Min.X, Min.Y, Min.Z);
    }

    public static BoundingBox CreateFromPoints(Span<Vector3> points)
    {
        Vector3 min = new(float.MaxValue);
        Vector3 max = new(float.MinValue);

        for (int i = 0; i < points.Length; ++i)
        {
            min = Vector3.Min(min, points[i]);
            max = Vector3.Max(max, points[i]);
        }

        return new BoundingBox(min, max);
    }

    public static BoundingBox CreateFromSphere(in BoundingSphere sphere)
    {
        return new(
            new Vector3(sphere.Center.X - sphere.Radius, sphere.Center.Y - sphere.Radius, sphere.Center.Z - sphere.Radius),
            new Vector3(sphere.Center.X + sphere.Radius, sphere.Center.Y + sphere.Radius, sphere.Center.Z + sphere.Radius)
            );
    }

    public static BoundingBox CreateMerged(in BoundingBox original, in BoundingBox additional)
    {
        return new(
            Vector3.Min(original.Min, additional.Min),
            Vector3.Max(original.Max, additional.Max)
        );
    }

    /// <summary>
    /// Transforms given <see cref="BoundingBox"/> using a given <see cref="Matrix4x4"/>.
    /// </summary>
    /// <param name="box">The source <see cref="BoundingBox"/>.</param>
    /// <param name="transform">A transformation matrix that might include translation, rotation, or uniform scaling.</param>
    /// <returns>The transformed BoundingBox.</returns>
    public static BoundingBox Transform(in BoundingBox box, in Matrix4x4 transform)
    {
        Transform(box, transform, out BoundingBox result);
        return result;
    }

    /// <summary>
    /// Transforms given <see cref="BoundingBox"/> using a given <see cref="Matrix4x4"/>.
    /// </summary>
    /// <param name="box">The source <see cref="BoundingBox"/>.</param>
    /// <param name="transform">A transformation matrix that might include translation, rotation, or uniform scaling.</param>
    /// <param name="result">The transformed BoundingBox.</param>
    public static void Transform(in BoundingBox box, in Matrix4x4 transform, out BoundingBox result)
    {
        Vector3 newCenter = Vector3.Transform(box.Center, transform);
        Vector3 oldEdge = box.Size * 0.5f;

        Vector3 newEdge = new(
            MathF.Abs(transform.M11) * oldEdge.X + MathF.Abs(transform.M12) * oldEdge.Y + MathF.Abs(transform.M13) * oldEdge.Z,
            MathF.Abs(transform.M21) * oldEdge.X + MathF.Abs(transform.M22) * oldEdge.Y + MathF.Abs(transform.M23) * oldEdge.Z,
            MathF.Abs(transform.M31) * oldEdge.X + MathF.Abs(transform.M32) * oldEdge.Y + MathF.Abs(transform.M33) * oldEdge.Z
        );

        result = new(newCenter - newEdge, newCenter + newEdge);
    }


    /// <inheritdoc/>
    public override readonly bool Equals([NotNullWhen(true)] object? obj) => (obj is BoundingBox other) && Equals(other);

    /// <summary>
    /// Determines whether the specified <see cref="BoundingBox"/> is equal to this instance.
    /// </summary>
    /// <param name="other">The <see cref="BoundingBox"/> to compare with this instance.</param>
    public readonly bool Equals(BoundingBox other)
    {
        return Min.Equals(other.Min)
            && Max.Equals(other.Max);
    }

    /// <summary>
    /// Compares two <see cref="BoundingBox"/> objects for equality.
    /// </summary>
    /// <param name="left">The <see cref="BoundingBox"/> on the left hand of the operand.</param>
    /// <param name="right">The <see cref="BoundingBox"/> on the right hand of the operand.</param>
    /// <returns>
    /// True if the current left is equal to the <paramref name="right"/> parameter; otherwise, false.
    /// </returns>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool operator ==(BoundingBox left, BoundingBox right)
    {
        return (left.Min == right.Min)
            && (left.Max == right.Max);
    }

    /// <summary>
    /// Compares two <see cref="BoundingBox"/> objects for inequality.
    /// </summary>
    /// <param name="left">The <see cref="BoundingBox"/> on the left hand of the operand.</param>
    /// <param name="right">The <see cref="BoundingBox"/> on the right hand of the operand.</param>
    /// <returns>
    /// True if the current left is unequal to the <paramref name="right"/> parameter; otherwise, false.
    /// </returns>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool operator !=(BoundingBox left, BoundingBox right)
    {
        return (left.Min != right.Min)
            || (left.Max != right.Max);
    }

    /// <inheritdoc/>
    public override readonly int GetHashCode() => HashCode.Combine(Min, Max);

    /// <inheritdoc />
    public override readonly string ToString() => ToString(format: null, formatProvider: null);

    /// <inheritdoc />
    public readonly string ToString(string? format, IFormatProvider? formatProvider)
    {
        return $"{{ {nameof(Min)} = {Min.ToString(format, formatProvider)}, {nameof(Max)} = {Max.ToString(format, formatProvider)} }}";
    }
}

File: Source/Alimer/Math/BoundingBox.cpp
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

#include "Alimer/Math/BoundingBox.h"
#include "Alimer/Math/BoundingSphere.h"
#include "Alimer/Math/Plane.h"
#include "Alimer/Core/Hash.h"

using namespace Alimer;

void BoundingBox::Set(const Vector3& min, const Vector3& max)
{
    this->min = min;
    this->max = max;
}

void BoundingBox::Set(float minX, float minY, float minZ, float maxX, float maxY, float maxZ)
{
    min.Set(minX, minY, minZ);
    max.Set(maxX, maxY, maxZ);
}

void BoundingBox::Set(const BoundingBox& box)
{
    min = box.min;
    max = box.max;
}

void BoundingBox::Set(const BoundingSphere& sphere)
{
    const Vector3& center = sphere.center;
    float radius = sphere.radius;

    // Calculate the minimum point for the box.
    min.x = center.x - radius;
    min.y = center.y - radius;
    min.z = center.z - radius;

    // Calculate the maximum point for the box.
    max.x = center.x + radius;
    max.y = center.y + radius;
    max.z = center.z + radius;
}

void BoundingBox::GetCorners(Vector3* corners) const
{
    ALIMER_ASSERT(corners);

    corners[0].Set(min.x, max.y, max.z);
    corners[1].Set(max.x, max.y, max.z);
    corners[2].Set(max.x, min.y, max.z);
    corners[3].Set(min.x, min.y, max.z);
    corners[4].Set(min.x, max.y, min.z);
    corners[5].Set(max.x, max.y, min.z);
    corners[6].Set(max.x, min.y, min.z);
    corners[7].Set(min.x, min.y, min.z);
}

BoundingBox BoundingBox::CreateFromPoints(const Vector3* points, size_t count) noexcept
{
    Vector3 min(FLT_MAX);
    Vector3 max(-FLT_MAX);

    for (size_t i = 0; i < count; ++i)
    {
        min = Vector3::Min(min, points[i]);
        max = Vector3::Max(max, points[i]);
    }

    return BoundingBox(min, max);
}

BoundingBox BoundingBox::CreateFromSphere(const BoundingSphere& sphere) noexcept
{
    return BoundingBox(
        Vector3(sphere.center.x - sphere.radius, sphere.center.y - sphere.radius, sphere.center.z - sphere.radius),
        Vector3(sphere.center.x + sphere.radius, sphere.center.y + sphere.radius, sphere.center.z + sphere.radius)
        );
}

BoundingBox BoundingBox::CreateMerged(const BoundingBox& original, const BoundingBox& additional) noexcept
{
    return BoundingBox(
        Vector3::Min(original.min, additional.min),
        Vector3::Max(original.max, additional.max)
        );
}

void BoundingBox::Transform(const BoundingBox& box, const Matrix4x4& transform, BoundingBox& result) noexcept
{
    Vector3 newCenter = Vector3::Transform(box.GetCenter(), transform);
    Vector3 oldEdge = box.GetSize() * 0.5f;

    Vector3 newEdge(
        Abs(transform.m11) * oldEdge.x + Abs(transform.m12) * oldEdge.y + Abs(transform.m13) * oldEdge.z,
        Abs(transform.m21) * oldEdge.x + Abs(transform.m22) * oldEdge.y + Abs(transform.m23) * oldEdge.z,
        Abs(transform.m31) * oldEdge.x + Abs(transform.m32) * oldEdge.y + Abs(transform.m33) * oldEdge.z
        );

    result.min = newCenter - newEdge;
    result.max = newCenter + newEdge;
}

BoundingBox BoundingBox::Transform(const BoundingBox& box, const Matrix4x4& transform) noexcept
{
    BoundingBox result;
    Transform(box, transform, result);
    return result;
}

ContainmentType BoundingBox::Contains(const Vector3& point) const noexcept
{
    if (min.x <= point.x && max.x >= point.x &&
        min.y <= point.y && max.y >= point.y &&
        min.z <= point.z && max.z >= point.z)
    {
        return ContainmentType::Contains;
    }

    return ContainmentType::Disjoint;
}

ContainmentType BoundingBox::Contains(const BoundingBox& box) const noexcept
{
    if (max.x < box.min.x || min.x > box.max.x)
        return ContainmentType::Disjoint;

    if (max.y < box.min.y || min.y > box.max.y)
        return ContainmentType::Disjoint;

    if (max.z < box.min.z || min.z > box.max.z)
        return ContainmentType::Disjoint;

    if (min.x <= box.min.x && (box.max.x <= max.x
        && min.y <= box.min.y && box.max.y <= max.y)
        && min.z <= box.min.z && box.max.z <= max.z)
    {
        return ContainmentType::Contains;
    }

    return ContainmentType::Intersects;
}

ContainmentType BoundingBox::Contains(const BoundingSphere& sphere) const noexcept
{
    Vector3 vector = Vector3::Clamp(sphere.center, min, max);
    float distance = Vector3::DistanceSquared(sphere.center, vector);

    if (distance > sphere.radius * sphere.radius)
        return ContainmentType::Disjoint;

    if (((min.x + sphere.radius <= sphere.center.x) && (sphere.center.x <= max.x - sphere.radius) && (max.x - min.x > sphere.radius))
        && ((min.y + sphere.radius <= sphere.center.y) && (sphere.center.y <= max.y - sphere.radius) && (max.y - min.y > sphere.radius))
        && ((min.z + sphere.radius <= sphere.center.z) && (sphere.center.z <= max.z - sphere.radius) && (max.z - min.z > sphere.radius)))
    {
        return ContainmentType::Contains;
    }

    return ContainmentType::Intersects;
}

bool BoundingBox::Intersects(const BoundingBox& box) const noexcept
{
    if (max.x < box.min.x || min.x > box.max.x)
    {
        return false;
    }
    if (max.y < box.min.y || min.y > box.max.y)
    {
        return false;
    }

    if (max.z < box.min.z || min.z > box.max.z)
    {
        return false;
    }

    return true;
}

bool BoundingBox::Intersects(const BoundingSphere& sphere) const noexcept
{
    return sphere.Intersects(*this);
}

bool BoundingBox::Intersects(const BoundingFrustum& frustum) const noexcept
{
    return frustum.Intersects(*this);
}

PlaneIntersectionType BoundingBox::Intersects(const Plane& plane) const noexcept
{
    // Calculate the distance from the center of the box to the plane.
    Vector3 center((min.x + max.x) * 0.5f, (min.y + max.y) * 0.5f, (min.z + max.z) * 0.5f);
    float distance = plane.Distance(center);

    // Get the extents of the box from its center along each axis.
    float extentX = (max.x - min.x) * 0.5f;
    float extentY = (max.y - min.y) * 0.5f;
    float extentZ = (max.z - min.z) * 0.5f;

    const Vector3& planeNormal = plane.normal;
    if (fabsf(distance) <= (fabsf(extentX * planeNormal.x) + fabsf(extentY * planeNormal.y) + fabsf(
        extentZ * planeNormal.z)))
    {
        return PlaneIntersectionType::Intersecting;
    }

    return (distance > 0.0f) ? PlaneIntersectionType::Front : PlaneIntersectionType::Back;
}

bool BoundingBox::Intersects(const Ray& ray, _Out_ float& distance) const noexcept
{
    // Intermediate calculation variables.
    float dnear = 0.0f;
    float dfar = 0.0f;
    float tmin = 0.0f;
    float tmax = 0.0f;

    const Vector3& origin = ray.position;
    const Vector3& direction = ray.direction;

    // X direction.
    float div = 1.0f / direction.x;
    if (div >= 0.0f)
    {
        tmin = (min.x - origin.x) * div;
        tmax = (max.x - origin.x) * div;
    }
    else
    {
        tmin = (max.x - origin.x) * div;
        tmax = (min.x - origin.x) * div;
    }
    dnear = tmin;
    dfar = tmax;

    // Check if the ray misses the box.
    if (dnear > dfar || dfar < 0.0f)
    {
        distance = 0.0f;
        return false;
    }

    // Y direction.
    div = 1.0f / direction.y;
    if (div >= 0.0f)
    {
        tmin = (min.y - origin.y) * div;
        tmax = (max.y - origin.y) * div;
    }
    else
    {
        tmin = (max.y - origin.y) * div;
        tmax = (min.y - origin.y) * div;
    }

    // Update the near and far intersection distances.
    if (tmin > dnear)
    {
        dnear = tmin;
    }
    if (tmax < dfar)
    {
        dfar = tmax;
    }
    // Check if the ray misses the box.
    if (dnear > dfar || dfar < 0.0f)
    {
        distance = 0.0f;
        return false;
    }

    // Z direction.
    div = 1.0f / direction.z;
    if (div >= 0.0f)
    {
        tmin = (min.z - origin.z) * div;
        tmax = (max.z - origin.z) * div;
    }
    else
    {
        tmin = (max.z - origin.z) * div;
        tmax = (min.z - origin.z) * div;
    }

    // Update the near and far intersection distances.
    if (tmin > dnear)
    {
        dnear = tmin;
    }
    if (tmax < dfar)
    {
        dfar = tmax;
    }

    // Check if the ray misses the box.
    if (dnear > dfar || dfar < 0.0f)
    {
        distance = 0.0f;
        return false;
    }

    // The ray intersects the box (and since the direction of a Ray is normalized, dnear is the distance to the ray).
    distance = dnear;
    return true;
}
float BoundingBox::GetWidth() const noexcept
{
    return GetHalfSize().x * 2.0f;
}

float BoundingBox::GetHeight() const noexcept
{
    return GetHalfSize().y * 2.0f;
}

float BoundingBox::GetDepth() const noexcept
{
    return GetHalfSize().z * 2.0f;
}

float BoundingBox::GetVolume() const noexcept
{
    Vector3 sides = max - min;
    return sides.x * sides.y * sides.z;
}

float BoundingBox::GetPerimeter() const noexcept
{
    Vector3 sides = max - min;
    return 4 * (sides.x + sides.y + sides.z);
}

float BoundingBox::GetSurfaceArea() const noexcept
{
    Vector3 sides = max - min;
    return 2 * (sides.x * sides.y + sides.x * sides.z + sides.y * sides.z);
}

bool BoundingBox::IsEmpty() const noexcept
{
    return min.x == max.x && min.y == max.y && min.z == max.z;
}

size_t BoundingBox::GetHashCode() const noexcept
{
    size_t hash = 0;
    HashCombine(hash, min, max);
    return hash;
}
